docs/advanced_documentation/algorithms/lu-solver.md

```diff
@@ -0,0 +1,714 @@
+section.
+
+### Power system equation properties
+
+#### Matrix properties of power system equations
+
+[Power flow equations](../../user_manual/calculations.md#power-flow-algorithms) are not Hermitian
+and also not positive (semi-)definite. As a result, methods that depend on these properties cannot
+be used.
+
+[State estimation equations](../../user_manual/calculations.md#state-estimation-algorithms) are
+intrinsically positive definite and Hermitian, but for
+[performance reasons](#performance-considerations), the matrix equation is augmented to achieve
+a consistent structure across the entire topology using Lagrange multipliers.
+
+#### Element size properties of power system equations
+
+Power system equations may contain elements of several orders of magnitude. This may lead to
+instabilities when solving the (matrix) equations due to non-linearly propagating (rounding) errors.
+Stability checks that limit those rounding errors need to function under these extreme conditions.
+
+### Block-sparsity considerations
+
+The power flow and state estimation equations involve block-sparse matrices: dense blocks, with a
+dimensionality varying between different calculation types, methods and symmetries, are distributed
+across the matrix in an extremely sparse way. The sparse structure can be pre-calculated, but the
+dense blocks need to be inverted separately. To make matters worse, the dense blocks may differ
+heavily in structure and contents between different nodes, and are often not solvable without
+pivoting.
+
+### Custom sparse LU-solver
+
+The choice for a custom LU-solver implementation comes from a number of considerations:
+
+* LU-decomposition is the best choice, because QR-decomposition and Cholesky-decomposition cannot
+  solve the power system equations efficiently as a consequence of the properties
+* Alternative LU-solver implementations are optimized for a variety of use cases that are less
+  sparse than the ones encountered in power systems.
+* Alternative LU-solver implementations do not have good block-sparse matrix equation solvers.
+
+## Implementation
+
+The LU-solver implemented in the power grid model consists of 3 components:
+
+* A block-sparse LU-solver that:
+  * handles factorization using the topological structure up to block-level
+  * solves the sparse matrix equation given the factorization
+* A dense LU-factor that handles individual blocks within the matrix equation
+
+### Dense LU-factorization
+
+The power grid model uses a modified version of the
+[`LUFullPiv` defined in Eigen](https://gitlab.com/libeigen/eigen/-/blob/3.4/Eigen/src/LU/FullPivLU.h)
+(credits go to the original author). The modification adds opt-in support for
+[pivot perturbation](#pivot-perturbation).
+
+#### Dense LU-factorization process
+
+The Gaussian elimination process itself is as usual. Let
+$M_p\equiv\begin{bmatrix}m_p && \vec{r}_p^T \\ \vec{q}_p && \hat{M}_p\end{bmatrix}$, where $p$ is
```

Comment:
This is for me an unusual way of describing the problem? Usually ,we are talking about 9 blocks in the process. But you only mentioned 4.

```
[
[a11, a12, a13],
[a21, a22, a23],
[a31, a32, a33],
]
```

Where
a11: block containing handled L/U combination where the pivots have been already handled
a12: already handled U part just above the current pivot
a13: already handled U part above the rest pivots to be handled
a21: already handled L part just in the left of current pivot
a22: current pivot
a23: current U part
a31: already handled L part in the left of the rest pivots to be handled
a32: current L part
a33: rest of the matrix which will be adjusted by current elimination process

Reply:
what do you mean i mentioned 4 blocks? i described the process for arbitrary block size. sym/asym is one thing, but the specific solvers can also add more dimensions to the block size. E.g. for NRPF, we treat the P and Q components separately, so it's not 3x3 but 9x9, right?

Since the process is iterative, you can step in at any pivot element $p$ and treat the rest as described until the next iteration $p$ needs to start.

Reply:
I mean, we should not omit the part which have been already factorized. We have to show the entire matrix in place, including the upper and left part of the matrix which already have the factorized L/U values.

Reply:
When you at pivot `k`, we should also show the rows and columns of `0 \to k-1`.

Reply:
i did it in a couple locations, but it makes the docs actually less readable the more i add it. can you please have a look to see whether it is sufficient as is?

Reply:
better?

Reply:
@TonyXiang8787 can you please double-check whether this is corrrectly resolved in the latest version?